app/slave/cluster_slave.py

```diff
@@ -56,6 +58,29 @@ def __init__(self, port, host, num_executors=10):
         self._build_teardown_coin = None
         self._base_executor_index = None
 
+        self._heartbeat_count = 0
+        self._heartbeat_count_threshold = Configuration['heartbeat_count_threshold']
+        self._heartbeat_frequency = Configuration['heartbeat_frequency']
+        self.scheduler = BackgroundScheduler()
+        self._heartbeat_job = None
+        self._logger.info('Heartbeat will run every {} seconds'.format(self._heartbeat_frequency))
+
+    def configure_heartbeat(self):
+        self._heartbeat_job = self.scheduler.add_job(self.heartbeat, 'interval', seconds=self._heartbeat_frequency)
+        return self.scheduler
+
+    def heartbeat(self):
+        state_url = self._master_api.url('slave', self._slave_id, 'heartbeat')
+        try:
+            self._network.post_with_digest(state_url, request_params={'slave': {'heartbeat': True}},
+                                           secret=Secret.get())
+            self._heartbeat_count = 0
+        except (requests.ConnectionError, requests.Timeout):
+            self._heartbeat_count += 1
+            if self._heartbeat_count >= self._heartbeat_count_threshold:
+                self._logger.warning('Master is not responding to heartbeats')
+                self._heartbeat_job.remove()
```

Comment:
Do you plan to do more here? It seems like this is a good point to shut down the slave service. This is related to my other comment about running in a SafeThread.

Reply:
On second thoughts, since slave is just going to sit idle after this point, it probably makes sense to just kill it. We can change the behavior in future.

Reply:
If that's the case then you should add a TODO here explaining the intention. Right now it looks incomplete (because it is)

Reply:
Also, how/why is the request to reconnect to the master different than the heartbeat request? Why can't you just keep sending requests to the master forever, regardless of if the master is responding (i.e. never call self._heartbeat_job.remove()).

When the master comes back online and the first successful heartbeat is received the master connects that slave.

You may have already explained this in your design doc. If you did, sorry for not noticing this until now.

Reply:
Interesting point. The TODO is to call connect_to_master for reconnection instead of just simple post to heartbeat when master is dead. The idea is that master's identity might have changed as it probably restarted or something. I guess in future heartbeat/connect_to_master can be combined and behave differently based on current state of connection, but we are probably not there yet. In any case, I'll clarify this in a TODO comment.